File: src/Remote/agent.py
from concurrent import futures

import grpc

import Remote.remote_pb2 as remote_pb2
import Remote.remote_pb2_grpc as remote_pb2_grpc

class RemoteAgent(remote_pb2_grpc.EnvironmentServicer):
	def __init__(self, agent):
		self.agent = agent

	def Initialize(self, attr, context):
		self.agent.set_param(attr.experiment)
		return remote_pb2.Empty()

	def Start(self, state, context):
		action = self.agent.start(state.values)
		return remote_pb2.Action(action = action)

	def Step(self, result, context):
		if result.terminal:
			self.agent.end(result.reward)
			return remote_pb2.Empty()

		(action, unused) = self.agent.step(result.reward, state.values)
		return remote_pb2.Action(action = action)



def serve(agent):
    server = grpc.server(futures.ThreadPoolExecutor(max_workers=10))
    remote_pb2_grpc.add_AgentServicer_to_server(
        RemoteAgent(agent), server)
    server.add_insecure_port('[::]:8081')
    server.start()
    server.wait_for_termination()

File: lib/remote/agent.go
package remote

import (
	"context"
	"sync"

	"github.com/stellentus/cartpoles/lib/logger"
	"github.com/stellentus/cartpoles/lib/rlglue"
)

type launcherAgent struct {
	client AgentClient
	ctx    context.Context
	wg     *sync.WaitGroup
	debug  logger.Debug
}

func newLauncherAgent(debug logger.Debug, ctx context.Context, wg *sync.WaitGroup) (*launcherAgent, error) {
	return &launcherAgent{
		ctx:   ctx,
		wg:    wg,
		debug: debug,
	}, nil
}

func (agent *launcherAgent) Initialize(run uint, experiment, environment rlglue.Attributes) error {
	port, err := getPort(2500+int(run), experiment)
	if err != nil {
		return err
	}

	cc, err := dialGrpc(agent.debug, port)
	if err != nil {
		return err
	}
	agent.client = NewAgentClient(cc)

	err = launchCommands(run, experiment, agent.ctx, agent.wg)
	if err != nil {
		return err
	}

	err = reattempt(func() error {
		_, err := agent.client.Initialize(agent.ctx, &AgentAttributes{
			Run:         &Run{Run: uint64(run)},
			Experiment:  &Attributes{Attributes: string(experiment)},
			Environment: &Attributes{Attributes: string(environment)},
		})
		return err
	})
	return err
}

func (agent *launcherAgent) Start(state rlglue.State) rlglue.Action {
	ctx := context.Background()
	action, _ := agent.client.Start(ctx, &State{Values: []float64(state)})
	return rlglue.Action(action.Action)
}

func (agent *launcherAgent) Step(state rlglue.State, reward float64) rlglue.Action {
	ctx := context.Background()
	action, _ := agent.client.Step(ctx, &StepResult{
		State:    &State{Values: []float64(state)},
		Reward:   reward,
		Terminal: false,
	})
	return rlglue.Action(action.Action)
}

func (agent *launcherAgent) End(state rlglue.State, reward float64) {
	ctx := context.Background()
	agent.client.Step(ctx, &StepResult{
		State:    &State{Values: []float64(state)},
		Reward:   reward,
		Terminal: true,
	})
}
